Userland/Libraries/LibCore/File.cpp

```diff
@@ -259,4 +266,213 @@ NonnullRefPtr<File> File::standard_error()
+        }
+        if (duplicate_count != 0) {
+            return copy_file_or_directory(get_duplicate_name(dst_path, duplicate_count), src_path);
+        }
+    }
+
+    auto source_or_error = File::open(src_path, IODevice::ReadOnly);
+    if (source_or_error.is_error())
+        return source_or_error.error();
+
+    auto& source = *source_or_error.value();
+
+    struct stat src_stat;
+    int rc = fstat(source.fd(), &src_stat);
+    if (rc < 0) {
+
+        return String(strerror(errno));
+    }
+
+    if (source.is_directory()) {
+        if (recursion_mode == RecursionMode::Disallowed)
+            return String("Tried to copy directory with RecursionMode::Disallowed");
+        return copy_directory(dst_path, src_path, src_stat);
+    }
+
+    if (link_mode == LinkMode::Allowed) {
+        if (link(src_path.characters(), dst_path.characters()) < 0)
+            return String(strerror(errno));
+
+        return {};
+    }
+
+    return copy_file(dst_path, src_stat, source);
+}
+
+Result<void, String> File::copy_file(const String& dst_path, const struct stat& src_stat, File& source)
+{
+
+    int dst_fd = creat(dst_path.characters(), 0666);
+    if (dst_fd < 0) {
+        if (errno != EISDIR)
+            return String(strerror(errno));
+
+        auto dst_dir_path = String::formatted("{}/{}", dst_path, LexicalPath(source.filename()).basename());
+        dst_fd = creat(dst_dir_path.characters(), 0666);
+        if (dst_fd < 0)
+            return String(strerror(errno));
+    }
+
+    ScopeGuard close_fd_guard([dst_fd]() { ::close(dst_fd); });
+
+    if (src_stat.st_size > 0) {
+        if (ftruncate(dst_fd, src_stat.st_size) < 0)
+            return String(strerror(errno));
+    }
+
+    for (;;) {
+        char buffer[32768];
+        ssize_t nread = ::read(source.fd(), buffer, sizeof(buffer));
+        if (nread < 0) {
+            return String(strerror(errno));
+        }
+        if (nread == 0)
+            break;
+        ssize_t remaining_to_write = nread;
+        char* bufptr = buffer;
+        while (remaining_to_write) {
+            ssize_t nwritten = ::write(dst_fd, bufptr, remaining_to_write);
+            if (nwritten < 0)
+                return String(strerror(errno));
+
+            ASSERT(nwritten > 0);
+            remaining_to_write -= nwritten;
+            bufptr += nwritten;
+        }
+    }
+
+    // NOTE: We don't copy the set-uid and set-gid bits.
+    auto my_umask = umask(0);
+    umask(my_umask);
+    int rc = fchmod(dst_fd, (src_stat.st_mode & ~my_umask) & ~06000);
+    if (rc < 0)
+        return String(strerror(errno));
+
+    return {};
+}
+
+Result<void, String> File::copy_directory(const String& dst_path, const String& src_path, const struct stat& src_stat, LinkMode link)
+{
+    int rc = mkdir(dst_path.characters(), 0755);
+    if (rc < 0)
+        return String(strerror(errno));
+
+    String src_rp = Core::File::real_path_for(src_path);
+    src_rp = String::format("%s/", src_rp.characters());
```

Comment:
Please don't use `String::format()` in new code as it is being removed. Use `String::formatted()` instead.